File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/middleware/gemini.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "../core/pressure_engine"
require_relative "../core/step_outcome"

module IAIso
  module Middleware
    # Google Gemini / Vertex AI middleware.
    module Gemini
      UsageMetadata = Data.define(:prompt_token_count, :candidates_token_count, :total_token_count) do
        def self.empty = new(prompt_token_count: 0, candidates_token_count: 0, total_token_count: 0)
      end
      Part = Data.define(:has_function_call) do
        def self.empty = new(has_function_call: false)
      end
      Candidate = Data.define(:parts) do
        def self.empty = new(parts: [])
      end
      Response = Data.define(:usage_metadata, :candidates)

      Options = Data.define(:raise_on_escalation) do
        def self.defaults = new(raise_on_escalation: false)
      end

      class BoundedModel
        # `raw` is duck-typed: must respond to `generate_content(request) -> Response`
        # and `model_name -> String`.
        def initialize(raw:, execution:, options: Options.defaults)
          @raw = raw
          @execution = execution
          @options = options
        end

        def generate_content(request)
          pre = @execution.check
          raise LockedError if pre == IAIso::Core::StepOutcome::LOCKED
          if pre == IAIso::Core::StepOutcome::ESCALATED && @options.raise_on_escalation
            raise EscalationRaisedError
          end
          resp =
            begin
              @raw.generate_content(request)
            rescue StandardError => e
              raise ProviderError, e.message
            end
          tokens = resp.usage_metadata.total_token_count.to_i
          if tokens.zero?
            tokens = resp.usage_metadata.prompt_token_count.to_i + resp.usage_metadata.candidates_token_count.to_i
          end
          tool_calls = 0
          resp.candidates.each do |c|
            c.parts.each { |p| tool_calls += 1 if p.has_function_call }
          end
          model = @raw.respond_to?(:model_name) && !@raw.model_name.to_s.empty? ? @raw.model_name : "unknown"
          @execution.record_step(IAIso::Core::StepInput.build(
            tokens: tokens, tool_calls: tool_calls,
            tag: "gemini.generateContent:#{model}",
          ))
          resp
        end
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-rust/crates/middleware/src/gemini.rs
//! Google Gemini / Vertex AI wrapper.

use crate::MiddlewareError;
use iaiso_core::{BoundedExecution, StepInput, StepOutcome};
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UsageMetadata {
    #[serde(default)]
    pub prompt_token_count: u64,
    #[serde(default)]
    pub candidates_token_count: u64,
    #[serde(default)]
    pub total_token_count: u64,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ContentPart {
    #[serde(default)]
    pub function_call: Option<Value>,
    #[serde(default)]
    pub text: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Content {
    #[serde(default)]
    pub parts: Vec<ContentPart>,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Candidate {
    #[serde(default)]
    pub content: Content,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Response {
    #[serde(default)]
    pub usage_metadata: UsageMetadata,
    #[serde(default)]
    pub candidates: Vec<Candidate>,
}

pub trait Model: Send + Sync {
    fn generate_content(&self, request: &Value) -> Result<Response, String>;
    fn model_name(&self) -> &str;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Options {
    pub raise_on_escalation: bool,
}

pub struct BoundedModel<'a, M: Model> {
    raw: M,
    execution: &'a BoundedExecution,
    opts: Options,
}

impl<'a, M: Model> BoundedModel<'a, M> {
    pub fn new(raw: M, execution: &'a BoundedExecution, opts: Options) -> Self {
        Self { raw, execution, opts }
    }

    pub fn generate_content(&self, request: &Value) -> Result<Response, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self
            .raw
            .generate_content(request)
            .map_err(MiddlewareError::Provider)?;
        let mut tokens = resp.usage_metadata.total_token_count;
        if tokens == 0 {
            tokens = resp.usage_metadata.prompt_token_count
                + resp.usage_metadata.candidates_token_count;
        }
        let tool_calls: u64 = resp
            .candidates
            .iter()
            .map(|c| {
                c.content
                    .parts
                    .iter()
                    .filter(|p| p.function_call.is_some())
                    .count() as u64
            })
            .sum();
        let model = self.raw.model_name().to_string();
        let _ = self.execution.record_step(StepInput {
            tokens,
            tool_calls,
            tag: Some(format!(
                "gemini.generateContent:{}",
                if model.is_empty() { "unknown" } else { &model }
            )),
            ..Default::default()
        });
        Ok(resp)
    }
}
